cli: actually drain after decommission

tbg opened the pull request:
I do not know how this was ever supposed to work, but the old code can not have
been intentional: it created a drain client but then did not consume from it.
This had the effect of kicking off the drain, but ~immediately cancelling the
context on the goroutine carrying it out on the decommissioning node. This PR
actually waits for the drain to complete.

There is a related issue here, though. You can't drain a node that isn't live,
so attempting to decommission a node that's down will fail on the drain step.
This is certainly true as of this PR, but should have been true before as well.

Our docs[^1] do not mention this rather large caveat at all, and it seems
strange anyway; if the node is down why would you let the failing drain get in
the way.  Really the code ought to distinguish between the case of a live and
dead node and react accordingly - this is not something this PR achieves.

Fixes https://github.com/cockroachdb/cockroach/issues/138265.
Fixes https://github.com/cockroachdb/cockroach/issues/137240.

[^1]: https://www.cockroachlabs.com/docs/v24.3/node-shutdown?filters=decommission#remove-nodes
Epic: none
Release note (ops change): the `node decommission` cli command now waits
until the target node is drained before marking it as fully
decommissioned. Previously, it would start drain but not wait, leaving
the target node briefly in a state where it would be unable to
communicate with the cluster but would still accept client requests
(which would then hang or hit unexpected errors).

tbg committed:
cli: actually drain after decommission

I do not know how this was ever supposed to work, but the old code can not have
been intentional: it created a drain client but then did not consume from it.
This had the effect of kicking off the drain, but ~immediately cancelling the
context on the goroutine carrying it out on the decommissioning node. This PR
actually waits for the drain to complete.

There is a related issue here, though. You can't drain a node that isn't live,
so attempting to decommission a node that's down will fail on the drain step.
This is certainly true as of this PR, but should have been true before as well.

Our docs[1] do not mention this rather large caveat at all, and it seems
strange anyway; if the node is down why would you let the failing drain get in
the way.  Really the code ought to distinguish between the case of a live and
dead node and react accordingly - this is not something this PR achieves.

Fixes https://github.com/cockroachdb/cockroach/issues/138265.

[1]: https://www.cockroachlabs.com/docs/v24.3/node-shutdown?filters=decommission#remove-nodes
Epic: none
Release note (ops change): the `node decommission` cli command now waits
until the target node is drained before marking it as fully
decommissioned. Previously, it would start drain but not wait, leaving
the target node briefly in a state where it would be unable to
communicate with the cluster but would still accept client requests
(which would then hang or hit unexpected errors).

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138732)
<!-- Reviewable:end -->

arulajmani reviewed (approved):
:shipit: 

Nice find!

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138732#-:-OGATZ4j9x2JTZvYMfkK:b-bvudbt)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kvoli and @tbg)

___
*[`-- commits` line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138732#-OGASe0E7s5twQ7g0wsv:-OGASe0E7s5twQ7g0wsw:byzevuk):*
Should we create an issue for this? I've seen the CC folks manually decommission dead nodes a few times, so I'd wager it's only a matter of time before someone quizzes us about this unexpected error. 


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138732)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @tbg)

___
*[`-- commits` line 17 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138732#-OGASe0E7s5twQ7g0wsv:-OGAW3vw4O4T78c6LyAM:bh0oz70):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we create an issue for this? I've seen the CC folks manually decommission dead nodes a few times, so I'd wager it's only a matter of time before someone quizzes us about this unexpected error.
</blockquote></details>

+1 on creating an issue, as described this appears like a glaring gap -- also surprised the test was only occasionally failing? Might be worth adding in another issue to beef up testcluster testing here.
___
*[`pkg/cli/node.go` line 617 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138732#-OGAXEOy3BLG1oJpfNki:-OGAXEOy3BLG1oJpfNkj:bs93oel) ([raw file](https://github.com/cockroachdb/cockroach/blob/acda747f16b15fee278a950fd5c46753a591b6c8/pkg/cli/node.go#L617)):*
> ```Go
> 				for {
> 					_, err := stream.Recv()
> 					if err == io.EOF {
> ```

Might be worth adding a comment above explaining the branching here on `io.EOF` for the happy case, since this code doesn't get too frequent attention.


<!-- Sent from Reviewable.io -->

stevendanna commented:
Holy moly.

tbg committed:
cli: improve logging

tbg committed:
cli: improve logging

tbg committed:
cli: fix typo

srosenberg commented:
Nice and so timely! I saw this failure [1] last night while smoke testing another PR, realizing it would probably make for an interesting finding–`n1` is decommissioned; immediately after, `/health?ready=1` is 200; then, we fail with `pq: server is not accepting clients, try another node`.


[1] https://teamcity.cockroachdb.com/repository/download/Cockroach_Nightlies_RoachtestNightlyGceBazel/18431348:id/decommission/mixed-versions/run_1/artifacts.zip!/test-post-assertions.log

tbg committed:
roachtest: small refactor

tbg committed:
roachtest: decommission/drains -> decommission/drains/{live,dead}

- rename `decommission/drains` to `decommission/drains/alive`
- add `decommission/drains/dead` flavor
- run these weekly instead of daily (we have enough other decom tests
  and since I'm adding one, we can also clamp down a bit)
- remove an old workaround that also accepted errors that
  should be impossible now that we properly wait for drain.

Epic: none

tbg commented:
Thanks for the reviews. I re-requested since I added some more commits: we now skip drain if target is not live. We also test this in the new /dead flavor of the roachtest. I've run both roachtests: they pass and drain only when supposed to.

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138732#-:-OGWCojoE5933oNE5nPg:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @kvoli and @tbg)



<!-- Sent from Reviewable.io -->

kvoli reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  :partying_face: 

> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138732#-:-OGWDF7a1aNPP_fg5DLB:b-hjeog3)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @arulajmani and @tbg)

___
*[`pkg/cmd/roachtest/tests/decommission.go` line 1075 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/138732#-OGWDqam0NNrwdXSRtSs:-OGWDqam0NNrwdXSRtSt:b-wm4m5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/a72cc4f5842c04873a5940acb80eec71326ccf66/pkg/cmd/roachtest/tests/decommission.go#L1075)):*
> ```Go
> 		// step is moot. If the target node is alive, the last decommission
> 		// invocation should have drained it, which we verify below.
> ```

nit: rm empty line above branching.


<!-- Sent from Reviewable.io -->

tbg commented:
TFTRs!

bors r+

tbg commented:
blathers backport 24.3

tbg commented:
(just testing)
blathers backport 24.1
